pilot/pkg/networking/core/v1alpha3/cluster_cache.go

```diff
@@ -178,10 +187,17 @@ func buildClusterKey(service *model.Service, port *model.Port, cb *ClusterBuilde
 		downstreamAuto:  cb.sidecarProxy() && port.Protocol.IsUnsupported(),
 		supportsIPv4:    cb.supportsIPv4,
 		service:         service,
-		destinationRule: proxy.SidecarScope.DestinationRule(model.TrafficDirectionOutbound, proxy, service.Hostname),
+		destinationRule: dr,
 		envoyFilterKeys: efKeys,
 		metadataCerts:   cb.metadataCerts,
 		peerAuthVersion: cb.req.Push.AuthnPolicies.GetVersion(),
 		serviceAccounts: cb.req.Push.ServiceAccounts(service.Hostname, service.Attributes.Namespace, port.Port),
+		endpointBuilder: endpoints.NewCDSEndpointBuilder(
```

Comment:
I think some of the slowdown might be that we do this unconditionally. None of our benchmarks are DNS resolution. This patch should have almost no impact.